File: 64-minimum-path-sum/minimum-path-sum.py
class Solution:
    def recursive(self,grid, r , c):
        if r < 0 or c < 0:
            return float('inf')
        if r == 0 and c == 0:
            return grid[0][0]
        #Recursive
        return grid[r][c] + min(self.recursive(grid, r - 1,c), self.recursive(grid, r, c - 1))
    
    def memoized(self,grid, r , c, dp):
        if r < 0 or c < 0:
            return float('inf')
        if r == 0 and c == 0:
            return grid[0][0]
        if dp[r][c] != -1:
            return dp[r][c]
        #Recursive
        dp[r][c] = grid[r][c] + min(self.memoized(grid, r - 1,c,dp), self.memoized(grid, r, c - 1,dp))
        return dp[r][c]
    
    def tabulation(self, grid):
        dp = [[0 for _ in range(len(grid[0]))] for _ in range(len(grid))]
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if r == 0 and c == 0:
                    dp[r][c] = grid[0][0]
                else:
                    dp[r][c] = grid[r][c] + min(dp[r - 1][c] if r > 0 else float('inf'), dp[r][c - 1] if c > 0 else float('inf'))
        return dp[len(grid) - 1][len(grid[0]) - 1]
    
    def optimized_tabulation(self, grid):
        prev = [0 for _ in range(len(grid[0]))]
        curr = [0 for _ in range(len(grid[0]))]
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if r == 0 and c == 0:
                    curr[c] = grid[0][0]
                else:
                    curr[c] = grid[r][c] + min(prev[c] if r > 0 else float('inf'), curr[c - 1] if c > 0 else float('inf'))
            prev = curr
            curr = [0 for _ in range(len(grid[0]))]
        return prev[len(grid[0]) - 1]
    
    def minPathSum(self, grid: List[List[int]]) -> int:
        return self.optimized_tabulation(grid)

File: 64-minimum-path-sum/minimum-path-sum.cpp
class Solution {
public:
    // // TC -> O(N * M * 2 ^ N)
    // // SC -> O(N + M)
    // int recursive(vector<vector<int>>& grid, int r, int c){
    //     if (r < 0 || c < 0){
    //         return 1000000000;
    //     }
    //     if (r == 0 && c == 0){
    //         return grid[r][c];
    //     }
    //     int left = grid[r][c] + recursive(grid, r, c - 1);
    //     int up = grid[r][c] + recursive(grid, r - 1 , c);
    //     return min(left, up);
    // }

    // // TC -> O(N * M * 2)
    // // SC -> O(N * M + (N + M))
    // int memoized(vector<vector<int>>& grid, vector<vector<int>>& dp, int r, int c){
    //     if (r < 0 || c < 0){
    //         return 100000000;
    //     }
    //     if (r == 0 and c == 0){
    //         return grid[r][c];
    //     }
    //     if (dp[r][c] != -1){
    //         return dp[r][c];
    //     }
    //     int left = grid[r][c] + memoized(grid, dp , r, c - 1);
    //     int up = grid[r][c] + memoized(grid, dp , r - 1 , c);
    //     dp[r][c] = min(left , up);
    //     return dp[r][c];
    // }

    // // TC -> O(N * M * 2)
    // // SC -> O(N * M)
    // int tabulation(vector<vector<int>>& grid){
    //     vector<vector<int>> dp(grid.size(), vector<int>(grid[0].size(), -1));
    //     for (int r = 0; r < grid.size(); r++){
    //         for (int c = 0; c < grid[0].size(); c++){
    //             if (r == 0 && c == 0){
    //                 dp[r][c] = grid[0][0];
    //                 continue;
    //             }
    //             int left = 100000000;
    //             int up = 1000000000;
    //             if (c - 1 >= 0){
    //                 left = dp[r][c - 1];
    //             }
    //             if (r - 1 >= 0){
    //                 up = dp[r - 1][c];
    //             }
    //             dp[r][c] = grid[r][c] + min(up, left);
    //         }
    //     }
    //     return dp[grid.size() - 1][grid[0].size() - 1];
    // }

    // TC -> O(N * M)
    // SC -> O(M)
    int optimizedTabulation(vector<vector<int>>& grid){
        vector<int> prev(grid[0].size(), -1);
        vector<int> curr(grid[0].size(), -1);
        for (int r = 0; r < grid.size(); r++){
            for(int c = 0; c < grid[0].size(); c++){
                if (r == 0 && c == 0){
                    curr[c] = grid[r][c];
                    continue;
                }
                int left = 100000000;
                int up = 100000000;
                if (c - 1 >= 0){
                    left = curr[c - 1];
                }
                if (r - 1 >= 0){
                    up = prev[c];
                }
                curr[c] = grid[r][c] + min(up, left);
            }
            prev = curr;
            curr = vector<int>(grid[0].size(), -1);
        }
        return prev[grid[0].size() - 1];
    }
    
    int minPathSum(vector<vector<int>>& grid) {
        return optimizedTabulation(grid);
    }
};